In [ ]:
LightGBM의 한 가지 단점으로 알려진것은 적은 데이터 세트에 적용할 경우 과적합이 발생하기 쉽다.

XGBoost는 LightGBM에 비해서 과적합 될 확률이 높지 않다.

training set을 training/validation으로 split을 한 이유는 무엇인가요?
-> 모델의 완성도를 좀 더 높이기 위해서이다.(여러번 검증함)
-> 과적합을 피하기 위해서.

In [ ]:
HyperOpt를 사용할 때 다음을 순서대로 나열하세요.
a.함수 설정 b.범위 설정 c.fmin()으로 x,y값 찾기
-> b,a,c

In [ ]:
다음중 랜덤 포레스트의 설명으로 옳지 않은 것은?

1. 가지를 나눌때마다 p개의 변수중 랜덤하게 m개(p보다 작음)의 변수만 고려한다.
2. 소수의 변수들에 대해 상관성이나 원소가 가진 영향력을 제거할 수 있다.
3. 원본 데이터의 건수가 200개일때 10개의 결정 트리 기반으로 학습하려고 하면
   20개씩 10개의 데이터 건수로 나눈다. (변수는 루트 데이터는 부트스트래핑으로 뽑아준다.)
4. 여러 개의 데이터세트르 분할 할 때, 부트스트래핑(bootstrapping) 분할 방식을 이용하며,
   이는 중첩(복원)을 허용한다.(데이터를 분할할때 부트스트래핑을 사용한다.)
-> 3

In [ ]:
hyperopt의 hp모듈을 사용하여 -20 - 20까지 2간격을 가지는 입력변수와
x와 -15 - 15까지 1간격으로 입력변수 y를 설정하세요.

from hyperopt import hp
search_space={'x':hp.quniform(-20,20,2),
             'y':hp.quniform(-15,15,1)}

In [ ]:
LightGBM이 XGBoost 대비 더 나은 점 두가지.
1. 학습시간이 더 빠르다.
2. 더 적은 메모리를 사용한다.

hyperopt를 사용할 때 목적함수의 변환값들을 어떠한 클래스 객체로 저장한다.
이때 사용하는 클래스는?
-> Trials

LightGBM은 최대한 균형 잡힌 트리를 유지하면서 분할하기 때문에 트리의 깊이가 최소화 될 수 있다.
-> X

HyperOpt 적용하여 최적의 파라미터 값을 찾을 때,
교차 검증 정확도의 평균에 -1을 곱하여 목적함수의 입력값(fmin()의 인수)으로 넣는 이유는 무엇인가?
-> 정확도는 높은 값일수록 좋은 성능을 지닌 지표라고 판단하는 반면,
   fmin() 함수는 값이 작을수록 더 좋은 최적화로 판단하기 때문이다.

In [ ]:
HyperOpt는 목적함수 반환 값의 최대값을 가지는 최적 입력값을 유추한다.
-> X

틀린 것을 모두 고르세요.
1. lightgbm은 xgboost보다 학습에 걸리는 시간이 훨씬 적다.
2. lightgbm은 xgboost와 예측 성능에 별다른 차이가 없다.
3.         은 적은 데이터 세트에 적용할 경우 과적합 위험이 크다.
4.         은 균형 트리 분할(level wise) 방식을 사용한다.
5.         은 다른 모델에 비해 단순하다는 장점이 있다.
-> 4,5

lightgbm의 파라미터 max_depth의 default값은 ?
max_depth값을 따로 지정하지 않을 경우 어떤 의미를 가지는가?
디폴트 -1. 트리의 깊이에 제한이 없다.

In [ ]:
사이킷런 분류 알고리즘에서 각 label데이터의 확률을 보여주는 메서드는 무엇인가?
-> predict_proba

ROC 곡선이 가운데 직선에 가까울수록 성능이 떨어지며,
멀어질수록 성능이 뛰어나다.
-> O

분류 결정 임곗값이 높아질수록 Positive로 예측할 확률이 높아진다.
-> X

In [ ]:
보기에서 틀린것을 고르세요.

1. 이진 분류 모델에서는 예측 확률이 큰 레이블 값으로 예측한다.
2. 재현율이 중요한 경우는 양성지표를 Negative로 잘못 판단시 큰 영향을 가지는 경우이다.
3. F1 스코어는 정밀도와 재현율을 결합한 지표이다.
4. F1 스코어는 정밀도와 재현율이 큰 차이를 가질수록 높은 값을 가진다.
5. FPR을 0으로 만들려면 임곗값을 1로 지정한다.
-> 4

보기에서 틀린것을 고르세요.
1. 정밀도가 중요한 경우는 음성데이터를 야성으로 잘못 판단하면 큰 영향이 발생하는 경우이다.
2. tp/(tp+fp)은 정밀도의 공식이다.
3. true positive rate은 재현율이다.
4. 전체 환자를 positive라 예측하고 실제 양성인 환자가 n명이라면 재현율은 0이 된다.
5. 전체 환자 중 1명만 positive라 예측하고 나머지를 전부 negative라 예측하면 정밀도는 1이 된다.
-> 4

오차 행렬의 원소값을 이용하여 정밀도, 재현율(민감도), 특이도에 대한 수식을 쓰시오.
정밀도= TP/(TP+FP)
재현율= TP/(TP+FN)   recall, sensitivity, TPR
특이도= TN/(TN+FP)

In [60]:
# 연료 종류(fl)를 레이블 값으로 설정해 이를 예측할 수 있는지 머신러닝을 구현해보자.
import pydataset
mpg = pydataset.data('mpg')
mpg.head()

manufacturer model  displ  year  cyl       trans drv  cty  hwy fl    class
1         audi    a4    1.8  1999    4    auto(l5)   f   18   29  p  compact
2         audi    a4    1.8  1999    4  manual(m5)   f   21   29  p  compact
3         audi    a4    2.0  2008    4  manual(m6)   f   20   31  p  compact
4         audi    a4    2.0  2008    4    auto(av)   f   21   30  p  compact
5         audi    a4    2.8  1999    6    auto(l5)   f   16   26  p  compact

In [61]:
from collections import Counter
Counter(mpg.fl)

Counter({'p': 52, 'r': 168, 'e': 8, 'd': 5, 'c': 1})

In [62]:
#연료 종류 중 p,r만 남기고 나머지는 삭제하세요.
import numpy as np

mpg['fl'].replace(['e','d','c'], np.nan, inplace=True)
mpg['fl'].dropna()

1      p
2      p
3      p
4      p
5      p
      ..
230    p
231    p
232    p
233    p
234    p
Name: fl, Length: 220, dtype: object

```
mpg.query("fl == 'p' | fl == 'r'") 이것도 같은 방법
```

In [63]:
#features와 labels를 나누어주세요.
feature = X = mpg.drop('fl',axis=1)
label = y = mpg[['fl']] 

In [64]:
feature

manufacturer   model  displ  year  cyl       trans drv  cty  hwy    class
1           audi      a4    1.8  1999    4    auto(l5)   f   18   29  compact
2           audi      a4    1.8  1999    4  manual(m5)   f   21   29  compact
3           audi      a4    2.0  2008    4  manual(m6)   f   20   31  compact
4           audi      a4    2.0  2008    4    auto(av)   f   21   30  compact
5           audi      a4    2.8  1999    6    auto(l5)   f   16   26  compact
..           ...     ...    ...   ...  ...         ...  ..  ...  ...      ...
230   volkswagen  passat    2.0  2008    4    auto(s6)   f   19   28  midsize
231   volkswagen  passat    2.0  2008    4  manual(m6)   f   21   29  midsize
232   volkswagen  passat    2.8  1999    6    auto(l5)   f   16   26  midsize
233   volkswagen  passat    2.8  1999    6  manual(m5)   f   18   26  midsize
234   volkswagen  passat    3.6  2008    6    auto(s6)   f   17   26  midsize

[234 rows x 10 columns]

In [65]:
#현재 피처값에 문자열이 존재한다. 이를 레이블 인코딩 하세요.
from sklearn.preprocessing import LabelEncoder

encoder = LabelEncoder()
label = encoder.fit_transform(label)
label

C:\Users\admin\anaconda3\lib\site-packages\sklearn\preprocessing\_label.py:115: DataConversionWarning: A column-vector y was passed when a 1d array was expected. Please change the shape of y to (n_samples, ), for example using ravel().
  y = column_or_1d(y, warn=True)


array([0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 1, 2, 1, 1,
       1, 0, 0, 0, 0, 0, 1, 2, 1, 2, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 2,
       1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 2, 1, 1, 1, 1, 2, 1, 1, 1, 1, 1, 2,
       1, 1, 1, 2, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1,
       1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 0, 1, 1, 1, 0, 1, 1, 1, 2, 0, 1, 1,
       1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 2, 1, 1, 1, 2, 1, 1, 0, 0, 1,
       1, 0, 1, 0, 1, 1, 1, 1, 1, 1, 1, 1, 1, 0, 0, 1, 1, 0, 1, 1, 0, 0,
       1, 0, 1, 1, 0, 1, 1, 1, 0, 1, 0, 1, 1, 1, 1, 0, 1, 0, 1, 1, 1, 1,
       1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1,
       1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 0, 0, 1, 2, 1, 1, 0, 0, 1, 1, 1,
       1, 2, 2, 1, 1, 1, 1, 0, 0, 0, 0, 0, 0, 0])

```
이것도 같은 방법

from sklearn.preprocessing import LabelEncoder

encoder = LabelEncoder()

columns = feature.columns

for column in ['manufacturer', 'model', 'trans', 'drv', 'class']:
    
    encoder.fit(feature[column])
    encoded = encoder.transform(feature[column])
    feature[column] = encoded

feature
```

In [66]:
# cross_val_score()을 사용해서 정제한 데이터를 구간을 5개로 나누고, 
# 평가 지표는 '정확도'를 사용하시오.
from sklearn.tree import DecisionTreeClassifier
from sklearn.model_selection import cross_val_score, cross_validate 

dt_clf = DecisionTreeClassifier()

scores = cross_val_score(dt_clf, features, label, scoring = 'accuracy', cv = 5)
scores

array([0.27659574, 0.44680851, 0.44680851, 0.46808511, 0.56521739])

In [67]:
# 교차검증
from sklearn.model_selection import cross_val_score
from sklearn.tree import DecisionTreeClassifier
import numpy as np

dt_clf = DecisionTreeClassifier()

scores = cross_val_score(dt_clf, features, label, scoring = 'accuracy', cv = 5)
print(f'정확도 : {np.mean(scores)}')

정확도 : 0.45337650323774276


# Credit card fraud detection

In [108]:
card = pd.read_csv('creditcard.csv')
card.head()

Time        V1        V2        V3        V4        V5        V6        V7  \
0   0.0 -1.359807 -0.072781  2.536347  1.378155 -0.338321  0.462388  0.239599   
1   0.0  1.191857  0.266151  0.166480  0.448154  0.060018 -0.082361 -0.078803   
2   1.0 -1.358354 -1.340163  1.773209  0.379780 -0.503198  1.800499  0.791461   
3   1.0 -0.966272 -0.185226  1.792993 -0.863291 -0.010309  1.247203  0.237609   
4   2.0 -1.158233  0.877737  1.548718  0.403034 -0.407193  0.095921  0.592941   

         V8        V9  ...       V21       V22       V23       V24       V25  \
0  0.098698  0.363787  ... -0.018307  0.277838 -0.110474  0.066928  0.128539   
1  0.085102 -0.255425  ... -0.225775 -0.638672  0.101288 -0.339846  0.167170   
2  0.247676 -1.514654  ...  0.247998  0.771679  0.909412 -0.689281 -0.327642   
3  0.377436 -1.387024  ... -0.108300  0.005274 -0.190321 -1.175575  0.647376   
4 -0.270533  0.817739  ... -0.009431  0.798278 -0.137458  0.141267 -0.206010   

        V26       V27       V28  Amount  Class  
0 -0.189115  0.133558 -0.021053  149.62      0  
1  0.125895 -0.008983  0.014724    2.69      0  
2 -0.139097 -0.055353 -0.059752  378.66      0  
3 -0.221929  0.062723  0.061458  123.50      0  
4  0.502292  0.219422  0.215153   69.99      0  

[5 rows x 31 columns]

In [83]:
card.shape

(284807, 31)

- LogisticRegression 모델 사용

In [84]:
# 1. Time 칼럼을 삭제하기
card = card.drop(['Time'], axis = 1)

In [92]:
# 2. 데이터셋 분리
from sklearn.model_selection  import train_test_split
features = card.iloc[:, :-1]
labels = card.iloc[:, -1]
X_train, X_test, y_train, y_test = train_test_split(features, labels, test_size = .3, stratify = labels) 

In [93]:
y_train.value_counts()/len(y_train) #len(y_train)은 전체값이다.

0    0.998275
1    0.001725
Name: Class, dtype: float64

In [95]:
# 3. 모델 생성하고 fit, predict, evaluate하기
from sklearn.linear_model import LogisticRegression
from sklearn.metrics import accuracy_score

lr_clf = LogisticRegression(max_iter = 1000) #반복개수를 1000번
lr_clf.fit(X_train, y_train)
preds = lr_clf.predict(X_test)
preds_proba = lr_clf.predict_proba(X_test)[:, 1]

In [96]:
from sklearn.metrics import roc_auc_score, confusion_matrix, accuracy_score, precision_score, recall_score, f1_score

def get_clf_eval(y_test, pred=None, pred_proba=None):
    confusion = confusion_matrix( y_test, pred)
    accuracy = accuracy_score(y_test , pred)
    precision = precision_score(y_test , pred)
    recall = recall_score(y_test , pred)
    f1 = f1_score(y_test,pred)
    # ROC-AUC 추가 
    roc_auc = roc_auc_score(y_test, pred_proba)
    print('오차 행렬')
    print(confusion)
    # ROC-AUC print 추가
    print('정확도: {0:.4f}, 정밀도: {1:.4f}, 재현율: {2:.4f},\
    F1: {3:.4f}, AUC:{4:.4f}'.format(accuracy, precision, recall, f1, roc_auc)) 

In [98]:
get_clf_eval(y_test, preds, preds_proba) #AUC 값을 내려고 preds_proba 시용함

오차 행렬
[[85286     9]
 [   57    91]]
정확도: 0.9992, 정밀도: 0.9100, 재현율: 0.6149,    F1: 0.7339, AUC:0.9832


여기까지가 LogisticRegression 모델을 사용한 것이다.

- LGBMClassifier 모델 사용

In [101]:
from lightgbm import LGBMClassifier

lgbm_clf =LGBMClassifier(n_estimators = 1000, num_leaves= 64, n_jobs= -1, boost_from_average= False) 
lgbm_clf.fit(X_train, y_train)
preds = lgbm_clf.predict(X_test)
preds_proba = lgbm_clf.predict_proba(X_test)[:, 1]

get_clf_eval(y_test, preds, preds_proba)

오차 행렬
[[85289     6]
 [   31   117]]
정확도: 0.9996, 정밀도: 0.9512, 재현율: 0.7905,    F1: 0.8635, AUC:0.9731


아까 91에서 117로 늘어났다.(높아졌다.)

# 데이터 정제(Scaling) 후 적용

In [102]:
card['Amount'].describe()

count    284807.000000
mean         88.349619
std         250.120109
min           0.000000
25%           5.600000
50%          22.000000
75%          77.165000
max       25691.160000
Name: Amount, dtype: float64

In [109]:
# 표준화 하기
from sklearn.preprocessing import StandardScaler

scaler = StandardScaler()
card['Amount'] = scaler.fit_transform(card['Amount'].values.reshape(-1,1))
card[:3]

Time        V1        V2        V3        V4        V5        V6        V7  \
0   0.0 -1.359807 -0.072781  2.536347  1.378155 -0.338321  0.462388  0.239599   
1   0.0  1.191857  0.266151  0.166480  0.448154  0.060018 -0.082361 -0.078803   
2   1.0 -1.358354 -1.340163  1.773209  0.379780 -0.503198  1.800499  0.791461   

         V8        V9  ...       V21       V22       V23       V24       V25  \
0  0.098698  0.363787  ... -0.018307  0.277838 -0.110474  0.066928  0.128539   
1  0.085102 -0.255425  ... -0.225775 -0.638672  0.101288 -0.339846  0.167170   
2  0.247676 -1.514654  ...  0.247998  0.771679  0.909412 -0.689281 -0.327642   

        V26       V27       V28    Amount  Class  
0 -0.189115  0.133558 -0.021053  0.244964      0  
1  0.125895 -0.008983  0.014724 -0.342475      0  
2 -0.139097 -0.055353 -0.059752  1.160686      0  

[3 rows x 31 columns]

In [115]:
# 표준화 후 결과

In [113]:
from sklearn.model_selection  import train_test_split
features = card.iloc[:, :-1]
labels = card.iloc[:, -1]
X_train, X_test, y_train, y_test = train_test_split(features, labels, test_size = .3, stratify = labels)

from sklearn.linear_model import LogisticRegression

lr_clf = LogisticRegression(max_iter = 1000) #반복개수를 1000번
lr_clf.fit(X_train, y_train)
preds = lr_clf.predict(X_test)
preds_proba = lr_clf.predict_proba(X_test)[:, 1]

get_clf_eval(y_test, preds, preds_proba)

오차 행렬
[[85280    15]
 [   67    81]]
정확도: 0.9990, 정밀도: 0.8438, 재현율: 0.5473,    F1: 0.6639, AUC:0.9387


In [114]:
from lightgbm import LGBMClassifier

lgbm_clf =LGBMClassifier(n_estimators = 1000, num_leaves= 64, n_jobs= -1, boost_from_average= False) 
lgbm_clf.fit(X_train, y_train)
preds = lgbm_clf.predict(X_test)
preds_proba = lgbm_clf.predict_proba(X_test)[:, 1]

get_clf_eval(y_test, preds, preds_proba)

오차 행렬
[[85288     7]
 [   27   121]]
정확도: 0.9996, 정밀도: 0.9453, 재현율: 0.8176,    F1: 0.8768, AUC:0.9818


# 데이터 정제(로그변환) 후 변환

In [116]:
card = pd.read_csv('creditcard.csv')

card = card.drop(['Time'], axis = 1)

In [118]:
# 로그변환 하기
card['Amount'] = np.log1p(card['Amount'])
card[:3]  #로그를 사용하는 이유: 숫자가 너무 커서

V1        V2        V3        V4        V5        V6        V7  \
0 -1.359807 -0.072781  2.536347  1.378155 -0.338321  0.462388  0.239599   
1  1.191857  0.266151  0.166480  0.448154  0.060018 -0.082361 -0.078803   
2 -1.358354 -1.340163  1.773209  0.379780 -0.503198  1.800499  0.791461   

         V8        V9       V10  ...       V21       V22       V23       V24  \
0  0.098698  0.363787  0.090794  ... -0.018307  0.277838 -0.110474  0.066928   
1  0.085102 -0.255425 -0.166974  ... -0.225775 -0.638672  0.101288 -0.339846   
2  0.247676 -1.514654  0.207643  ...  0.247998  0.771679  0.909412 -0.689281   

        V25       V26       V27       V28    Amount  Class  
0  0.128539 -0.189115  0.133558 -0.021053  1.794216      0  
1  0.167170  0.125895 -0.008983  0.014724  0.835352      0  
2 -0.327642 -0.139097 -0.055353 -0.059752  1.937197      0  

[3 rows x 30 columns]

In [119]:
from sklearn.model_selection  import train_test_split
features = card.iloc[:, :-1]
labels = card.iloc[:, -1]
X_train, X_test, y_train, y_test = train_test_split(features, labels, test_size = .3, stratify = labels)

from sklearn.linear_model import LogisticRegression

lr_clf = LogisticRegression(max_iter = 1000)
lr_clf.fit(X_train, y_train)
preds = lr_clf.predict(X_test)
preds_proba = lr_clf.predict_proba(X_test)[:, 1]

print(get_clf_eval(y_test, preds, preds_proba))
print()


from lightgbm import LGBMClassifier

lgbm_clf =LGBMClassifier(n_estimators = 1000, num_leaves= 64, n_jobs= -1, boost_from_average= False) 
lgbm_clf.fit(X_train, y_train)
preds = lgbm_clf.predict(X_test)
preds_proba = lgbm_clf.predict_proba(X_test)[:, 1]

print(get_clf_eval(y_test, preds, preds_proba))

오차 행렬
[[85280    15]
 [   58    90]]
정확도: 0.9991, 정밀도: 0.8571, 재현율: 0.6081,    F1: 0.7115, AUC:0.9827
None

오차 행렬
[[85295     0]
 [   29   119]]
정확도: 0.9997, 정밀도: 1.0000, 재현율: 0.8041,    F1: 0.8914, AUC:0.9812
None


# 이상치 데이터 제거 후 모델 학습 /예측 /평가

In [ ]:
# 1. 이상치 찾기
IQR = 3/4 - 1/4

3/4 + 1.5IQR
1/4 - 1.5IQR

In [120]:
card.Class.value_counts()

0    284315
1       492
Name: Class, dtype: int64

In [133]:
# 조건에 만족하는 것을 뽑아내었다.
fraud = card[card['Class'] == 1]
fraud

V1        V2        V3        V4        V5        V6        V7  \
541    -2.312227  1.951992 -1.609851  3.997906 -0.522188 -1.426545 -2.537387   
623    -3.043541 -3.157307  1.088463  2.288644  1.359805 -1.064823  0.325574   
4920   -2.303350  1.759247 -0.359745  2.330243 -0.821628 -0.075788  0.562320   
6108   -4.397974  1.358367 -2.592844  2.679787 -1.128131 -1.706536 -3.496197   
6329    1.234235  3.019740 -4.304597  4.732795  3.624201 -1.357746  1.713445   
...          ...       ...       ...       ...       ...       ...       ...   
279863 -1.927883  1.125653 -4.518331  1.749293 -1.566487 -2.010494 -0.882850   
280143  1.378559  1.289381 -5.004247  1.411850  0.442581 -1.326536 -1.413170   
280149 -0.676143  1.126366 -2.213700  0.468308 -1.120541 -0.003346 -2.234739   
281144 -3.113832  0.585864 -5.399730  1.817092 -0.840618 -2.943548 -2.208002   
281674  1.991976  0.158476 -2.583441  0.408670  1.151147 -0.096695  0.223050   

              V8        V9       V10  ...       V21       V22       V23  \
541     1.391657 -2.770089 -2.772272  ...  0.517232 -0.035049 -0.465211   
623    -0.067794 -0.270953 -0.838587  ...  0.661696  0.435477  1.375966   
4920   -0.399147 -0.238253 -1.525412  ... -0.294166 -0.932391  0.172726   
6108   -0.248778 -0.247768 -4.801637  ...  0.573574  0.176968 -0.436207   
6329   -0.496358 -1.282858 -2.447469  ... -0.379068 -0.704181 -0.656805   
...          ...       ...       ...  ...       ...       ...       ...   
279863  0.697211 -2.064945 -5.587794  ...  0.778584 -0.319189  0.639419   
280143  0.248525 -1.127396 -3.232153  ...  0.370612  0.028234 -0.145640   
280149  1.210158 -0.652250 -3.463891  ...  0.751826  0.834108  0.190944   
281144  1.058733 -1.632333 -5.245984  ...  0.583276 -0.269209 -0.456108   
281674 -0.068384  0.577829 -0.888722  ... -0.164350 -0.295135 -0.072173   

             V24       V25       V26       V27       V28    Amount  Class  
541     0.320198  0.044519  0.177840  0.261145 -0.143276  0.000000      1  
623    -0.293803  0.279798 -0.145362 -0.252773  0.035764  1.984152      1  
4920   -0.087330 -0.156114 -0.542628  0.039566 -0.153029  1.869416      1  
6108   -0.053502  0.252405 -0.657488 -0.827136  0.849573  1.628131      1  
6329   -1.632653  1.488901  0.566797 -0.010016  0.146793  0.526589      1  
...          ...       ...       ...       ...       ...       ...    ...  
279863 -0.294885  0.537503  0.788395  0.292680  0.147968  1.941430      1  
280143 -0.081049  0.521875  0.739467  0.389152  0.186637  0.448086      1  
280149  0.032070 -0.739695  0.471111  0.385107  0.194361  1.680466      1  
281144 -0.183659 -0.328168  0.606116  0.884876 -0.253700  1.872622      1  
281674 -0.450261  0.313267 -0.289617  0.002988 -0.015309  1.563069      1  

[492 rows x 30 columns]

In [134]:
# 여기서 이상치 찾아내기
fraud['V14']

541      -4.289254
623      -1.692029
4920     -1.470102
6108     -6.771097
6329     -6.079337
            ...   
279863   -6.665177
280143   -5.210141
280149   -4.057162
281144   -6.416628
281674   -1.948883
Name: V14, Length: 492, dtype: float64

In [139]:
pct25 = np.percentile(fraud.V14, 25)
pct75 = np.percentile(fraud.V14, 75)

IQR = pct75 - pct25

upper_out = pct75 + 1.5 * IQR
under_out = pct25 - 1.5 * IQR

cond1 = fraud['V14'] >= upper_out
cond2 = fraud['V14'] <= under_out
fraud.V14[cond1 | cond2] 

8296   -19.214325
8615   -18.822087
9035   -18.493773
9252   -18.049998
Name: V14, dtype: float64

In [142]:
outlier = fraud[(fraud['V14'] > upper_out) | (fraud['V14'] < under_out)]
outlier.index

Int64Index([8296, 8615, 9035, 9252], dtype='int64')

In [149]:
card.drop(outlier.index, axis=0, inplace=True)

In [151]:
card[:3]

V1        V2        V3        V4        V5        V6        V7  \
0 -1.359807 -0.072781  2.536347  1.378155 -0.338321  0.462388  0.239599   
1  1.191857  0.266151  0.166480  0.448154  0.060018 -0.082361 -0.078803   
2 -1.358354 -1.340163  1.773209  0.379780 -0.503198  1.800499  0.791461   

         V8        V9       V10  ...       V21       V22       V23       V24  \
0  0.098698  0.363787  0.090794  ... -0.018307  0.277838 -0.110474  0.066928   
1  0.085102 -0.255425 -0.166974  ... -0.225775 -0.638672  0.101288 -0.339846   
2  0.247676 -1.514654  0.207643  ...  0.247998  0.771679  0.909412 -0.689281   

        V25       V26       V27       V28    Amount  Class  
0  0.128539 -0.189115  0.133558 -0.021053  1.794216      0  
1  0.167170  0.125895 -0.008983  0.014724  0.835352      0  
2 -0.327642 -0.139097 -0.055353 -0.059752  1.937197      0  

[3 rows x 30 columns]

In [152]:
from sklearn.model_selection  import train_test_split
features = card.iloc[:, :-1]
labels = card.iloc[:, -1]
X_train, X_test, y_train, y_test = train_test_split(features, labels, test_size = .3, stratify = labels)

from sklearn.linear_model import LogisticRegression

lr_clf = LogisticRegression(max_iter = 1000)
lr_clf.fit(X_train, y_train)
preds = lr_clf.predict(X_test)
preds_proba = lr_clf.predict_proba(X_test)[:, 1]

print(get_clf_eval(y_test, preds, preds_proba))
print()


from lightgbm import LGBMClassifier

lgbm_clf =LGBMClassifier(n_estimators = 1000, num_leaves= 64, n_jobs= -1, boost_from_average= False) 
lgbm_clf.fit(X_train, y_train)
preds = lgbm_clf.predict(X_test)
preds_proba = lgbm_clf.predict_proba(X_test)[:, 1]

print(get_clf_eval(y_test, preds, preds_proba))

오차 행렬
[[85274    21]
 [   53    93]]
정확도: 0.9991, 정밀도: 0.8158, 재현율: 0.6370,    F1: 0.7154, AUC:0.9755
None

오차 행렬
[[85290     5]
 [   31   115]]
정확도: 0.9996, 정밀도: 0.9583, 재현율: 0.7877,    F1: 0.8647, AUC:0.9802
None


# SMOTE로 오버샘플링

In [153]:
from imblearn.over_sampling import SMOTE

In [157]:
# 데이터셋 분리
features = card.iloc[:, :-1]
labels = card.iloc[:, -1]
X_train, X_test, y_train, y_test = train_test_split(features, labels, test_size = .3, stratify = labels)

# SMOTE 적용
smote = SMOTE(random_state=0)
X_train, y_train = smote.fit_resample(X_train, y_train) #y_train을 oversampling해주었다.

# 모델 생성1
lr_clf = LogisticRegression(max_iter = 1000)
lr_clf.fit(X_train, y_train)
preds = lr_clf.predict(X_test)
preds_proba = lr_clf.predict_proba(X_test)[:, 1]

print(get_clf_eval(y_test, preds, preds_proba))
print()

#모델 생성2
lgbm_clf =LGBMClassifier(n_estimators = 1000, num_leaves= 64, n_jobs= -1, boost_from_average= False) 
lgbm_clf.fit(X_train, y_train)
preds = lgbm_clf.predict(X_test)
preds_proba = lgbm_clf.predict_proba(X_test)[:, 1]

print(get_clf_eval(y_test, preds, preds_proba))

오차 행렬
[[83095  2200]
 [   11   135]]
정확도: 0.9741, 정밀도: 0.0578, 재현율: 0.9247,    F1: 0.1088, AUC:0.9794
None

오차 행렬
[[85278    17]
 [   31   115]]
정확도: 0.9994, 정밀도: 0.8712, 재현율: 0.7877,    F1: 0.8273, AUC:0.9797
None


false가 많아졌는데 이유는 oversampling을 했기 때문이다.